### PR TITLE
Fixes #15543 - External method to search fact's hostgroup

### PR DESCRIPTION
**Pull request by shlomizadok**

### CODE REVIEW

File: app/models/fact_value.rb [hunk 1]
```diff
@@ -45,6 +45,15 @@ def self.search_by_host(key, operator, value)
     { :conditions => "fact_values.id IN(#{search.join(',')})" }
   end
 
+  def self.search_by_hostgroup(key, operator, value)
+    search_term = value =~ /\A\d+\Z/ ? 'id' : 'name'
+    conditions = sanitize_sql_for_conditions(["hostgroups.#{search_term} #{operator} ?", value_to_sql(operator, value)])
+    search     = FactValue.joins(:hostgroup).where(conditions).uniq.pluck('fact_values.id')
+
+    return { :conditions => "1=0" } if search.empty?
+    { :conditions => "fact_values.id IN(#{search.join(',')})" }
```
Comment (ohadlevy) [inline app/models/fact_value.rb]: why not use an inner query? if search is a scope, it will be done automatically.

Comment (tbrisker) [reply app/models/fact_value.rb]: Even better, I think this should work just as well and get rid of the extra query completely:

``` ruby
search_term = value =~ /\A\d+\Z/ ? 'id' : 'name'
conditions = sanitize_sql_for_conditions(["hostgroups.#{search_term} #{operator} ?", value_to_sql(operator, value)])
{ :joins => :hostgroup, :conditions => conditions }
```

Comment (shlomizadok) [reply app/models/fact_value.rb]: @tbrisker - this works perfect. Thanks !

**Comment by tbrisker** [comment]
@shlomizadok Could you apply the same refactoring to the host search above while at it?

**Comment by shlomizadok** [comment]
@tbrisker - yeah, tried to DRY them as well.

File: app/models/fact_value.rb [hunk 2]
```diff
@@ -37,12 +37,17 @@ class FactValue < ActiveRecord::Base
...
+    search_by('host', key, operator, value)
+  end
 
-    return { :conditions => "1=0" } if search.empty?
-    { :conditions => "fact_values.id IN(#{search.join(',')})" }
+  def self.search_by_hostgroup(key, operator, value)
+    search_by('hostgroup', key, operator, value)
+  end
+
+  def self.search_by(host_or_hg, key, operator, value)
```
Comment (tbrisker) [inline app/models/fact_value.rb]: you can even DRY this more- key is either `host` or `host.hostgroup`, so you can use the same method for both and just check the key to decide which table to use.

**Review by tbrisker — CHANGES_REQUESTED** [review]

**Comment by shlomizadok** [comment]
@tbrisker - got it as DRY as I can...

File: app/models/fact_value.rb [hunk 3]
```diff
@@ -35,14 +35,12 @@ class FactValue < ActiveRecord::Base
   scope :root_only, -> { with_roots.where(:fact_names => {:ancestry => nil}) }
 
   validates :fact_name_id, :uniqueness => { :scope => :host_id }
-
-  def self.search_by_host(key, operator, value)
+
```
[the remark this answers is not in the document]
Comment (shlomizadok) [reply app/models/fact_value.rb]: @houndci-bot woof!

**Review by tbrisker — APPROVED** [review]

File: app/models/fact_value.rb (hunk 2 above)
Comment (shlomizadok) [reply app/models/fact_value.rb]: Excellent idea. Changed :+1:

**Comment by shlomizadok** [comment]
Foreman [test] failure seem unrelated

**Comment by shlomizadok** [comment]
@theforeman/core - Any clue why this is failing?

**Comment by tbrisker** [comment]
@shlomizadok this is the cause of the failures: http://projects.theforeman.org/issues/16599

**Comment by dLobatog** [comment]
[test foreman]

**Comment by shlomizadok** [comment]
[test foreman]

**Comment by tbrisker** [comment]
Merged as 5ce4c4c51917fdcf377fb168157a099108e5557f, thanks @shlomizadok!

**Pull request closed without merging**